### DISCUSSION TITLE
Can You Export to Code?

### DISCUSSION THREAD

**Discussion by websmyths** [langflow-ai/langflow#26] [Q&A]
Is there a way to export a workflow to Python or Nodejs code? Or to import the JSON into a Python or node project?

**Comment by ogabrielluiz** [comment]
You can but this is something that requires a lot more testing. 

The buttons to import/export the flows are located in the bottom left of the canvas, close to the zoom buttons.
@anovazzi1 Maybe we should revise the location and design of these buttons to make them easier to find. What do you think?

Once you have exported, you can import into python doing something close to this:
https://github.com/logspace-ai/langflow/blob/472b0d657d8b2c086d4a0f47cb326dc43dad5404/tests/test_loading.py

This is still subject to change but it won't be too different from this.

**Reply by websmyths** [reply]
Thanks! I will give it a try.

**Comment by kinthaiofficial** [comment]
Yes — Langflow supports exporting to Python code, which is one of the more valuable features for moving from prototype to production.

The flow is:
1. Build your pipeline visually in Langflow
2. Click the `</>` (Code) button or Export → Python
3. Get a Python file with the equivalent LangChain chain

Example of what the export looks like for a RAG pipeline:

```python
# Exported from Langflow
from langchain.chains import RetrievalQA
from langchain_community.vectorstores import Chroma
from langchain_openai import ChatOpenAI, OpenAIEmbeddings

def build_chain():
    embeddings = OpenAIEmbeddings(model="text-embedding-3-small")
    vectorstore = Chroma(
        persist_directory="./chroma_db",
        embedding_function=embeddings,
    )
    
    llm = ChatOpenAI(model="gpt-4o", temperature=0)
    
    chain = RetrievalQA.from_chain_type(
        llm=llm,
        chain_type="stuff",
        retriever=vectorstore.as_retriever(search_kwargs={"k": 4}),
        return_source_documents=True,
    )
    return chain
```

The export is most useful as a starting point — you'll likely need to adjust:
- Error handling
- Async support (Langflow export tends to be sync)
- Environment-specific config (hardcoded values → env vars)
- Production logging/observability

The gap between the visual prototype and the production code is smaller than you'd expect — Langflow's internal representation maps closely to LangChain's Python API since that's what it's built on.

Is there a specific node type or pipeline you're trying to export? Some custom components may not export cleanly.
